agent/utils/buffer-ring/bigqueue/bigqueue.go

```diff
@@ -0,0 +1,223 @@
+		drainChan: make(chan int64),
+	}
+	out.drainRunner()
+	return out, nil
+}
+
+// Push stores agent message in cache.
+func (r *Ring) Push(msg *agentpb.AgentMessage) {
+	out, err := proto.Marshal(msg)
+	if err != nil {
+		r.l.Errorf("marshal proto while inserting message to cache: %+v", err)
+		return
+	}
+	if size := int64(len(out)) + indexEntrySize; size > r.totalSize {
+		r.l.Errorf("data size: '%d' overflows free cache space: '%d'", size, r.totalSize)
+		return
+	}
+	r.m.Lock()
+	_, err = r.fq.Enqueue(out)
+	r.m.Unlock()
+	if err != nil {
+		r.l.Errorf("inserting to cache: %+v", err)
+		return
+	}
+	if amount := r.usedSize() + drainThreshold - r.totalSize; amount > 0 {
+		r.requestDrain(amount)
+	}
+}
+
+// Send unmarshals agent message and send it in loop until dries the cache.
+func (r *Ring) Send(send func(msg *agentpb.AgentMessage) error) {
+	defer r.runGC()
+	for {
+		r.m.Lock()
+		idx, b, err := r.fq.Dequeue()
+		r.m.Unlock()
+		if err != nil {
+			r.l.Errorf("reading entry from cache: %+v", err)
+			return
+		}
+		if idx == -1 {
+			return
+		}
+		var out agentpb.AgentMessage
+		if err := proto.Unmarshal(b, &out); err != nil {
+			r.l.Errorf("unmarshal entry from cache: %+v", err)
+			return
+		}
+		if err := send(&out); err != nil {
+			r.l.Errorf("sending entry from cache: %+v", err)
+			return
+		}
+	}
+}
+
+// Close closes cache database.
+func (r *Ring) Close() error {
+	select {
+	case <-r.done:
+	default:
+		close(r.done)
+		<-r.drainDone
+		return r.fq.Close()
+	}
+	return nil
+}
+
+// initPaths creates all paths for queue to use. Original repo creates directories with perm error.
+func initPaths(dir string) error {
+	for _, path := range []string{
+		"",
+		bigqueue.IndexFileName,
+		bigqueue.DataFileName,
+		bigqueue.MetaFileName,
+		bigqueue.FrontFileName,
+	} {
+		if err := os.MkdirAll(filepath.Join(dir, path), os.ModePerm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// drain tries to drain cache.
+func (r *Ring) drain(amount int64) error {
+	defer r.runGC()
+	for size := int64(0); size < amount; {
+		r.m.Lock()
+		idx, b, err := r.fq.Dequeue()
+		r.m.Unlock()
+		if err != nil {
+			return err
+		}
+		if idx == -1 {
+			return nil
+		}
+		size += int64(len(b)) + indexEntrySize
+	}
+	return nil
+}
+
+func (r *Ring) usedSize() int64 {
+	r.m.Lock()
+	status := r.fq.Status()
+	r.m.Unlock()
+	sum := int64(metaFileSize)
+	for _, list := range status.IndexFileList {
+		sum += list.Size
+	}
+	for _, list := range status.DataFileList {
+		sum += list.Size
+	}
+	return sum
+}
+
+func (r *Ring) runGC() {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if err := r.fq.Gc(); err != nil {
+		r.l.Errorf("run gc: %+v", err)
+	}
+}
+
+func (r *Ring) requestDrain(amount int64) {
+	select {
+	case r.drainChan <- amount:
+	default:
+	}
+}
+
+func (r *Ring) drainRunner() {
+	go func() {
+		for {
+			select {
+			case <-r.done:
+				close(r.drainDone)
+				return
+			case amount := <-r.drainChan:
+				if err := r.drain(amount); err != nil {
```

Comment:
We need to recalculate the `amount` value here.

Reply:
ok